src/lib.rs

```diff
@@ -811,6 +862,41 @@ pub trait Perform {
     /// subsequent characters were ignored.
     fn esc_dispatch(&mut self, _intermediates: &[u8], _ignore: bool, _byte: u8) {}
 
+    /// Invoked when the beginning of a new SOS (Start of String) sequence is
+    /// encountered.
+    fn sos_start(&mut self) {}
+
+    /// Invoked when the beginning of a new APC (Application Program Command)
+    /// sequence is encountered.
+    fn apc_start(&mut self) {}
+
+    /// Invoked when the beginning of a new PM (Privacy Message) sequence is
+    /// encountered.
+    fn pm_start(&mut self) {}
+
+    /// Invoked for every valid character (0x20-0xFF) in a SOS (Start of String)
+    /// sequence.
+    fn sos_dispatch(&mut self, _byte: u8) {}
+
+    /// Invoked for every valid character (0x20-0xFF) in an APC (Application
+    /// Program Command) sequence.
+    fn apc_dispatch(&mut self, _byte: u8) {}
+
+    /// Invoked for every valid character (0x20-0xFF) in a PM (Privacy Message)
+    /// sequence.
+    fn pm_dispatch(&mut self, _byte: u8) {}
```

Comment:
I'm not sure the naming `_dispatch` is consistent with the rest of our methods. The existing `_dispatch` functions generally represent the dispatch of an entire escape sequence in full, while this function just represents the dispatch of a single byte in the sequence.

I think this is much closer to the way esc works in VTE, where it is split in `hook`, `put`, and `unhook`.

Other parts of this patch already make use of the `_put` nomenclature, so I think it's better to be consistent and rename these functions to `sos/pm/apc_put`.

Reply:
You're absolutely right, considering the method on the internal `OpaqueDispatch` trait is called `opaque_put` and then it ends up calling `sos/apc/pm_dispatch()`...